File: wip_tests/arrayiterator.py
#!/usr/bin/env python

class ArrayIterator():
    def __init__(self, size_x, size_y, mode):
        self.size_x = size_x
        self.size_y = size_y
        self.mode = mode
        self.current = 0
        if mode == "spiral":
            self.current_length = 0
            self.length = 1
            self.current_x = (size_x-1) / 2
            self.current_y = (size_y-1) / 2
            self.dx = 1
            self.dy = 0
            self.iterations = 0
    def next(self):
        x = None
        if self.current >= self.size_x * self.size_y:
            return None, None
        if self.mode == "topleft":
            x = self.current % self.size_x
            y = self.current / self.size_x
        elif self.mode == "topright":
            x = self.size_x - (self.current % self.size_x) - 1
            y = self.current / self.size_x
        elif self.mode == "bottomleft":
            x = self.current % self.size_x
            y = self.size_y - (self.current / self.size_x) - 1
        elif self.mode == "bottomright":
            x = self.size_x - (self.current % self.size_x) - 1
            y = self.size_y - (self.current / self.size_x) - 1
        elif self.mode == "spiral":
            while x is None:
                if (self.current_x>=0) and (self.current_x<self.size_x) and (self.current_y>=0) and (self.current_y<self.size_y):
                    x = self.current_x
                    y = self.current_y
                if self.current_length == self.length:
                    # we're at a corner
                    self.iterations += 1
                    self.current_length = 0
                    # switch direction
                    self.dx, self.dy = -self.dy, self.dx
                if self.iterations == 2:
                    self.length += 1
                    self.iterations = 0
                self.current_length += 1
                self.current_x += self.dx
                self.current_y += self.dy
        self.current += 1
        return x, y

import sys

def main():
    size_x = int(sys.argv[1])
    size_y = int(sys.argv[2])
    mode = sys.argv[3]
    iterator = ArrayIterator(size_x, size_y, mode)
    current = 0
    array = [0] * (size_x*size_y)
    while True:
        x, y = iterator.next()
        if x is None:
            break
        current += 1
        array[y*size_x+x] = current
    for y in range(size_y):
        for x in range(size_x):
            print "%d\t" % array[y*size_x+x],
        print ""

main()

File: wip_tests/arrayiterator.cpp
/*

***********************************************************
* © Benoit Leveau, inc. All rights reserved. *
***********************************************************

*/

#include <cstddef>
#include <utility>
#include <iostream>

#include "arrayiterator.hpp"

int main(int argc, char **argv)
{
    size_t size_x = atoi(argv[1]);
    size_t size_y = atoi(argv[2]);
    std::string mode = argv[3];
    ArrayIterationMode iter_mode;
    if (mode == "topleft")
        iter_mode = ArrayIterationMode::TOPLEFT;
    else if (mode == "topright")
        iter_mode = ArrayIterationMode::TOPRIGHT;
    else if (mode == "bottomleft")
        iter_mode = ArrayIterationMode::BOTTOMLEFT;
    else if (mode == "bottomright")
        iter_mode = ArrayIterationMode::BOTTOMRIGHT;
    else if (mode == "spiral")
        iter_mode = ArrayIterationMode::SPIRAL;
    ArrayIterator it(size_x, size_y, iter_mode);
    size_t *array = new size_t[size_x*size_y];
    size_t current = 0;
    while (it.valid)
    {
        auto pair = it.next();
        size_t x = pair.first;
        size_t y = pair.second;
        array[y*size_x+x] = ++current;
    }
    for (size_t y=0; y<size_y; ++y)
    {
        for (size_t x=0; x<size_x; ++x)
        {
            std::cout << array[y*size_x+x] << "\t";
        }
        std::cout << std::endl;
    }
    
    return 0;
}
